src/mumble/widgets/TrayIcon.cpp

```diff
@@ -0,0 +1,204 @@
+// Copyright 2024 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#include "TrayIcon.h"
+
+#include "../ClientUser.h"
+#include "../MainWindow.h"
+#include "../Global.h"
+
+#include <QApplication>
+
+TrayIcon::TrayIcon() : QSystemTrayIcon(Global::get().mw), m_statusIcon(Global::get().mw->qiIcon) {
+	applyIcon(m_statusIcon);
+
+	setToolTip("Mumble");
+
+	assert(Global::get().mw);
+
+	QObject::connect(Global::get().mw, &MainWindow::talkingStatusChanged, this, &TrayIcon::on_icon_update);
+	QObject::connect(Global::get().mw, &MainWindow::disconnectedFromServer, this, &TrayIcon::on_icon_update);
+
+	QObject::connect(Global::get().mw, &MainWindow::windowHighlighted, this, &TrayIcon::on_timer_triggered);
+	QObject::connect(Global::get().mw, &MainWindow::windowActivated, this, &TrayIcon::on_tray_unhighlight);
+
+	m_highlightTimer = new QTimer(this);
+	m_highlightTimer->setSingleShot(true);
+	QObject::connect(m_highlightTimer, &QTimer::timeout, this, &TrayIcon::on_timer_triggered);
+
+	QObject::connect(this, &QSystemTrayIcon::activated, this, &TrayIcon::on_icon_clicked);
+
+	// messageClicked is buggy in Qt on some platforms and we can not do anything about this (QTBUG-87329)
+	QObject::connect(this, &QSystemTrayIcon::messageClicked, this, &TrayIcon::on_showAction_triggered);
+
+	m_showAction = new QAction(tr("Show"), Global::get().mw);
+	QObject::connect(m_showAction, &QAction::triggered, this, &TrayIcon::on_showAction_triggered);
+
+	m_hideAction = new QAction(tr("Hide"), Global::get().mw);
+	QObject::connect(m_hideAction, &QAction::triggered, this, &TrayIcon::on_hideAction_triggered);
+
+	QObject::connect(Global::get().mw->qaTalkingUIToggle, &QAction::triggered, this, &TrayIcon::updateContextMenu);
+
+	m_contextMenu = new QMenu(Global::get().mw);
+	QObject::connect(m_contextMenu, &QMenu::aboutToShow, this, &TrayIcon::updateContextMenu);
+
+	// Some window managers hate it when a tray icon sets an empty context menu...
+	updateContextMenu();
+
+	setContextMenu(m_contextMenu);
+
+	show();
+}
+
+void TrayIcon::on_icon_update() {
+	std::reference_wrapper< QIcon > newIcon = Global::get().mw->qiIcon;
```

Comment:
why are you using a `reference_wrapper` for this? That class is usually only used when storing references as a member or in a container.